(Contributor Recognition Infrastructure) Milestone2.3: Add contributor badges with the mobile view

chris7716 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of n/a.
2. This PR does the following: Added badges for contributors: Doc https://docs.google.com/document/d/16SciF9u-AngOdmvBaSyjYCAYrppR6lTE-S2vwhhh86A/edit3. 

- This is stacked on https://github.com/oppia/oppia/pull/16289
- This branch is taken from the milestone2.2 branch, which is also in-review currently. So only the following files are relevant for this PR.

```
assets/images/contributor_dashboard/*
core/templates/app.constants.ts
core/templates/pages/contributor-dashboard-page/contributor-badge/*
core/templates/pages/contributor-dashboard-page/question-badge/*
core/templates/pages/contributor-dashboard-page/translation-badge/*
```


## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The linter/Karma presubmit checks have passed locally on your machine.
- [ ] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [ ] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

https://user-images.githubusercontent.com/41409951/199207545-92a2e14c-9d2f-4684-8564-95c68712186e.mov


https://user-images.githubusercontent.com/41409951/199207758-969658da-0414-470e-842a-b2822a3b8101.mov

![image](https://user-images.githubusercontent.com/41409951/199209605-8b280795-6bb3-4f46-b11f-578f2cf2507c.png)

![image](https://user-images.githubusercontent.com/41409951/199209696-6dcd2e6e-ffd4-45dc-9da1-8839dbadc616.png)



#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->


https://user-images.githubusercontent.com/41409951/199207328-5802dc2c-4425-4423-8edf-9abe3085ac66.mov



#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

chris7716 committed:
service file name changed to backend-api

chris7716 committed:
backend api file added

chris7716 committed:
previous service file removed from CODEOWNERS

chris7716 committed:
new backend service file added to  CODEOWNERS

chris7716 committed:
Merge remote-tracking branch 'upstream/develop' into develop
Merged with uostream

chris7716 committed:
Merge remote-tracking branch 'upstream/develop' into develop
merged with develop

chris7716 committed:
Merge remote-tracking branch 'upstream/develop' into develop
uptodated

chris7716 committed:
Merge branch 'oppia:develop' into develop

chris7716 committed:
Merge branch 'oppia:develop' into develop

chris7716 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop
Update branch

chris7716 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop
Update branch

chris7716 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop
Update local

chris7716 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop
Update local

chris7716 committed:
First pass

chris7716 committed:
First pass

chris7716 committed:
Finalize views

chris7716 committed:
Add stats component

chris7716 committed:
Resolve conflicts and refactor new component

chris7716 committed:
Cover tests for stats component

chris7716 committed:
Improve mobile views

chris7716 committed:
Remove mock responses

chris7716 committed:
Remove unwanted spaces

chris7716 committed:
Add question stats update

chris7716 committed:
First pass

chris7716 committed:
Add minor changes

sagangwee commented:
@chris7716 Noting that the doc link in the description doesn't seem to be working.

Some initial thoughts:
- Can we make the "Translations" header bottom margin match the "Questions" header, i.e. closer to the badges box? Like it is in the mocks: https://www.figma.com/file/SNOHAatFcCvdvIAdxAzF6E/Contributor-Dashboard-Achievements?node-id=1124%3A3686
- Can we left-align the badges like in the mocks? They appear to be centered here.
- Can we increase the spacing between the badge headers and the badges to more closely align with the mocks?
- Can we increase the font size of the "Language" filter label and make the filter box styling consistent with the other contributor dashboard filter boxes, e.g. making the border gray instead of black (same goes for the badge box)? Also, I think the filter box can be made less wide a bit.

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.html:
```diff
@@ -0,0 +1,523 @@
+        <div class="dropdown-label">Badge Type</div>
+        <div #mobileBadgeTypeDropdown class="oppia-contribution-stats-type-selector-mobile-container e2e-test-topic-selector">
+            <div class="oppia-stats-type-selector-inner-container e2e-test-topic-selector-selected" (click)="toggleMobileBadgeTypeDropdown()">
+              {{mobileTranslationBadgesShown ? 'Translation' : 'Question'}}
```
nit: `showMobileTranslationBadges` would read better.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.html:
```diff
@@ -0,0 +1,523 @@
+<div class="stats-container">
+    <div class="stats-contribution-type-mobile-container">
```
Seems odd that the mobile container class is used here, but it's always displayed.

chris7716 replied:
In desktop view it is set as none. Only when the width < 700px it is displayed using css class

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.html:
```diff
@@ -0,0 +1,523 @@
+            </div>
+        </div>
+    </div>
+    <div class="non-mobile-badge-containier">
```
Typo in "containier". Also, it may be cleaner to drop "non-mobile" from the class name.

chris7716 replied:
Done and renamed

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.html:
```diff
@@ -0,0 +1,523 @@
+    <div class="oppia-lazy-loading-container"
+         [hidden]="!dataLoading">
+      <lazy-loading></lazy-loading>
+      <lazy-loading></lazy-loading>
```
Is it possible to make the loading box larger to more closely reflect the end-state?

chris7716 replied:
Is this sufficient? 
![image](https://user-images.githubusercontent.com/41409951/201453627-ce07c062-f5ba-4d77-836d-08545f5d9669.png)

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.html:
```diff
@@ -0,0 +1,523 @@
+      <div class="oppia-contributions-required-message">
+        Start contributing to see your badges!
+      </div>
+    </mat-card>
```
Do these cover the case if the user has question badges, but no translation badges on desktop?

chris7716 replied:
This will be only shown if there are no translation badges AND the user has no question rights(that means no question badges). If the user has question rights but has not achieved any badges yet, but he has disabled badges.

Since user can submit translations for any language, we will have to create disabled badges for all of them if we want to show them. I do not think it's necessary and will be time consuming to created disabled badges for all languages. Disabled badges for translation submissions will be shown only if the user has review right for the particular language.

But for the mobile view it's different because Badge type dropdown item for Questions only visible if the user has rights for question submission/reviews.

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.spec.ts:
```diff
@@ -0,0 +1,303 @@
+      ],
+      schemas: [NO_ERRORS_SCHEMA]
+    }).compileComponents();
+  }));
```
Add new line before and after.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.spec.ts:
```diff
@@ -0,0 +1,303 @@
+      expect(component.mobileTranslationBadgesShown).toBeTrue();
+    }));
+
+    it('should show show question badges type in mobile', fakeAsync(() => {
```
Fix the double words, e.g. "should should".

Also, I'm not sure these test names match the parent "when user navigates to contributor badge page", since these require user action.

chris7716 replied:
Done. Added the user action as "when user clicks on it"

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.spec.ts:
```diff
@@ -0,0 +1,303 @@
+          return fixture.debugElement.nativeElement.querySelector(
+            '.oppia-stats-type-selector-dropdown-container');
+        };
+      });
```
Add new line.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.spec.ts:
```diff
@@ -0,0 +1,303 @@
+          component.onDocumentClick(fakeClickAwayEvent);
+          fixture.detectChanges();
+          expect(component.dropdownShown).toBe(false);
+          expect(getDropdownOptionsContainer()).toBeFalsy();
```
Remember to add new lines for the test structure.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.html:
```diff
@@ -0,0 +1,523 @@
+                    [ngClass]="{'oppia-stats-type-selector-dropdown-option-selected' : mobileTranslationBadgesShown}">
+                  Translation
+              </div>
+              <div *ngIf="userCanSuggestQuestions || userCanReviewQuestionSuggestions" class="oppia-stats-type-selector-dropdown-option e2e-test-topic-selector-option"
```
We seem to be using the `userCanSuggestQuestions || userCanReviewQuestionSuggestions` logic a lot in this file, so it may be worth declaring a `userHasQuestionRights` variable in the component.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,395 @@
+})
+export class ContributorBadgeComponent {
+  @ViewChild('dropdown', {'static': false}) dropdownRef!: ElementRef;
+  @ViewChild(
```
Maybe add new line to be consistent with below.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,395 @@
+  userCanReviewQuestionSuggestions: boolean;
+  userCanSuggestQuestions: boolean;
+  reviewableLanguages = [];
+  constructor(
```
Add new line here.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,395 @@
+  text: string;
+  language?: string;
+  badgeUnlocked: boolean;
+}
```
Add new line.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,395 @@
+  value: number;
+  text: string;
+  language?: string;
+  badgeUnlocked: boolean;
```
Since this is the context of a badge already, we can drop "badge". Also, since it's boolean, `isUnlocked` would be better.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,395 @@
+}
+
+interface Badge {
+  value: number;
```
Value seems too generic. Maybe something like "contributionCount".

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,395 @@
+import { UserService } from 'services/user.service';
+import { ContributionAndReviewStatsService } from '../services/contribution-and-review-stats.service';
+
+interface TotalContributionsAndReviews {
```
`ContributionCounts`?

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,395 @@
+            allCorrectionBadgesCreated: false
+          };
+        } else {
+          this.totalTranslationStats[languageDescription].submissions += (
```
```suggestion
          this.totalTranslationStats[languageDescription].submissions = (
```

I don't think you need to add the value, just set it. Here and below.

chris7716 replied:
Here one stat record gives the number of submissions for a given topic in a given language. If we want to get the total submissions for the language we need to get the sum of contributions for all topics so we need to add the value rather setting the value.

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,395 @@
+    }
+
+    for (let language in this.totalTranslationStats) {
+      const predefinedTranslationBadges = this.getPredefinedBadges(
```
I think the below logic can be simplified quite a bit if we had a function like `getObtainedBadges(contributionCount, contributionType, contributionSubType, language?)`, which we would call for each type of badge/language. That way we wouldn't need the "get predefined/non-predefined" or "all badges created" functions. The function could look like:

```
badges = []
for level in initial badge levels:
  if  contributionCount >= level:
    add badge
  else:
    add locked badge
    return

level = 1000
while contributionCount >= level:
  same logic as above...
```

You can also call this function above for each stat count above, so you wouldn't need the `TotalContributionsAndReviews` objects to store intermediate values.

chris7716 replied:
Changed the logic.

But I think still we need to have `TotalContributionsAndReviews` since before creating badges, we need to have total submissions, reviews, corrections(per language if translations) and one backend dict represents stats per a topic.

sagangwee commented on core/templates/pages/contributor-dashboard-page/translation-badge/translation-badge.component.spec.ts:
```diff
@@ -0,0 +1,128 @@
+import { TranslationBadgeComponent } from './translation-badge.component';
+import { AppConstants } from 'app.constants';
+
+describe('Translation bdge component', () => {
```
Typo.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/question-badge/question-badge.component.spec.ts:
```diff
@@ -0,0 +1,100 @@
+import { AppConstants } from 'app.constants';
+import { QuestionBadgeComponent } from './question-badge.component';
+
+describe('Question bdge component', () => {
```
Typo.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,395 @@
+    };
+    if (language) {
+      badge.language = language;
+    }
```
I think we can just assign the language above as well, since if language is undefined, it should be the same.

chris7716 replied:
Done

sagangwee reviewed (changes requested):
@chris7716 Thanks, took a pass.

chris7716 committed:
Address comments

chris7716 committed:
Address changes

chris7716 committed:
Rollback unwanted changes

chris7716 reviewed (commented):
@sagangwee PTAL Thanks!



https://user-images.githubusercontent.com/41409951/201467511-c836ae7f-be79-4f9a-9907-aef0183d8028.mov

chris7716 committed:
Address comments

chris7716 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into milestone2.2
Update

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,300 @@
+    return badges;
+  }
+
+  createBadgeObject(
+      contributionCount: number,
+      text: string,
+      isUnlocked: boolean,
+      language?: string): Badge {
+    const badge: Badge = {
+      contributionCount,
+      text,
+      isUnlocked,
+      language
+    };
+
+    return badge;
+  }
```
This function doesn't seem to really do much. I would just get rid of it.

chris7716 replied:
Removed

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,300 @@
+            value, contributionSubType, true, language)
+        );
+      } else {
+        badges.push(
```
Maybe add a comment here explaining that you're adding a locked badge.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,300 @@
+          this.createBadgeObject(
+            (value - contributionCount), contributionSubType, false, language)
+        );
+        break;
```
Should be able to return here.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,300 @@
+      contributionCount: number,
+      contributionSubType: string, language?: string): Badge[] {
+    const badges: Badge[] = [];
+    for (const value of AppConstants.CONTRIBUTOR_BADGE_INITIAL_LEVELS) {
```
```suggestion
    for (const level of AppConstants.CONTRIBUTOR_BADGE_INITIAL_LEVELS) {
```
Seems more descriptive.

chris7716 replied:
Done

sagangwee commented:
@chris7716 Can you post a screencast of badges obtained past the defined levels?

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,300 @@
+          this.createBadgeObject(
+            (level - contributionCount), contributionSubType, false, language)
+        );
+        break;
```
Can return early instead.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.html:
```diff
@@ -0,0 +1,552 @@
+        <div *ngIf="!dataLoading && (languages.length > 0)">
+            <mat-card layout="row" class="list-view-item oppia-contributions-required-message-container">
+                <div class="badge-box">
+                    <div>Submission</div>
```
These labels should be plural like in the mocks.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.html:
```diff
@@ -0,0 +1,552 @@
+            </div>
+        </div>
+    </div>
+    <mat-card [hidden]="dataLoading" *ngIf="(showMobileTranslationBadges && !translationBadges[selectedLanguage])" layout="row" class="contributions-required-container mobile-badge-container list-view-item">
+      <div class="oppia-contributions-required-message">
+        Start contributing to see your badges!
+      </div>
+    </mat-card>
+    <mat-card [hidden]="dataLoading" *ngIf="!translationBadges[selectedLanguage] && !userHasQuestionRights" layout="row" class="contributions-required-container desktop-badge-container list-view-item">
+      <div class="oppia-contributions-required-message">
+        Start contributing to see your badges!
+      </div>
+    </mat-card>
```
Can we add tests for these two cases? Could you also post screenshots of these two cases as well?

chris7716 replied:
When has no translation badges AND no question rights(no question badges).

![image](https://user-images.githubusercontent.com/41409951/202260337-c01e3902-44f7-4744-9a07-f37055a2c07f.png)

When has no translation badges AND has question rights(has unlocked question badges)
![image](https://user-images.githubusercontent.com/41409951/202260737-e8e41ccd-995f-4ab3-a05b-300b1ba80972.png)

When has translation badges AND no question rights(no question badges)
![image](https://user-images.githubusercontent.com/41409951/202261442-359df1f2-9a55-4ec6-a3ec-94879d02fdc5.png)

Updated test cases

sagangwee commented on core/templates/app.constants.ts:
```diff
@@ -206,5 +206,25 @@ export const AppConstants = {
+  CONTRIBUTION_STATS_SUBTYPE_REVIEW: 'review',
+  CONTRIBUTION_STATS_SUBTYPE_CORRECTION: 'correction',
+  CONTRIBUTION_STATS_TYPES: {
+    TRANSALTION_CONTRIBUTION: {
+      NAME: 'translationContribution',
+      DISPLAY_NAME: 'Translation Contributions'
+    },
+    TRANSALTION_REVIEW: {
```
Noting the typos.

sagangwee commented on core/templates/pages/contributor-dashboard-page/question-badge/question-badge.component.html:
```diff
@@ -0,0 +1,41 @@
+        {{contributionTypeText}}
+    </div>
+</div>
+<div *ngIf="isUnlocked === 'false'"  
```
```suggestion
<div *ngIf="!isUnlocked"  
```

Should work since `isUnlocked` is a boolean. Here and elsewhere.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/question-badge/question-badge.component.html:
```diff
@@ -0,0 +1,41 @@
+<div *ngIf="isUnlocked === 'true'" class="badge-container">
```
```suggestion
<div *ngIf="isUnlocked" class="badge-container">
```

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/translation-badge/translation-badge.component.html:
```diff
@@ -0,0 +1,55 @@
+    </div>
+</div>
+<div *ngIf="isUnlocked === 'false'"  
+     ngbTooltip="{{value}} more {{contributionTypeText.toLowerCase()}} to achieve this"
```
```suggestion
     ngbTooltip="{{value}} more {{contributionTypeText.toLowerCase()}} to achieve this badge"
```

and elsewhere.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,300 @@
+          stat.accepted_translations_with_reviewer_edits_count);
+      });
+    }
+    if (response.question_contribution_stats.length > 0) {
+      response.question_contribution_stats.map((stat) => {
+        this.totalQuestionStats.submissions = stat.accepted_questions_count;
+      });
+    }
+    if (response.question_review_stats.length > 0) {
+      response.question_review_stats.map((stat) => {
+        this.totalQuestionStats.reviews = stat.reviewed_questions_count;
+        this.totalQuestionStats.corrections = (
+          stat.accepted_questions_with_reviewer_edits_count);
+      });
+    }
```
Don't you also need to add these counts since there is one question stat per topic? Also, there should be unit tests verifying these counts/obtained badges.

chris7716 replied:
Yes done.

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,300 @@
+      this.userCanSuggestQuestions ||
+      this.userCanReviewQuestionSuggestions);
+
+    const response = await this.contributionAndReviewStatsService.fetchAllStats(
```
```suggestion
    const allContributionStats = await this.contributionAndReviewStatsService.fetchAllStats(
```

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/question-badge/question-badge.component.ts:
```diff
@@ -0,0 +1,34 @@
+})
+export class QuestionBadgeComponent {
+  @Input() type!: string;
+  @Input() value: number;
```
```suggestion
  @Input() contributionCount: number;
```

Ditto for translation badge.

chris7716 replied:
Done

sagangwee reviewed (changes requested):
@chris7716 Thanks, took a pass. Also, remember to assign your reviewers so that it shows up in our dashboard.

chris7716 committed:
AAddress comments

chris7716 committed:
Address comments

chris7716 committed:
Remove language from PageableStats

chris7716 committed:
Address more comments

chris7716 committed:
Merge branch 'milestone2.2' of https://github.com/chris7716/oppia into milestone2.3
Update branch

chris7716 committed:
Address more comments

chris7716 reviewed (commented):
@sagangwee PTAL

chris7716 committed:
Remove unwanted changes

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.spec.ts:
```diff
@@ -0,0 +1,350 @@
+          component.ngOnInit();
+        }));
+
+        it('should create translation badges', fakeAsync(() => {
```
This test name doesn't seem to match the assertions.

chris7716 replied:
Updated

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.spec.ts:
```diff
@@ -0,0 +1,350 @@
+        expect(component.selectedLanguage).toBe('Hindi');
+      }));
+
+      it('should show show translation badges type in mobile', fakeAsync(() => {
```
```suggestion
      it('should show translation badges type in mobile', fakeAsync(() => {
```

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.spec.ts:
```diff
@@ -0,0 +1,350 @@
+        expect(component.showMobileTranslationBadges).toBeTrue();
+      }));
+
+      it('should show show question badges type in mobile', fakeAsync(() => {
```
```suggestion
      it('should show question badges type in mobile', fakeAsync(() => {
```

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,304 @@
+    this.mobileDropdownShown = false;
+  }
+
+  selectBadgeType(showMobileTranslationBadges: boolean): void {
```
```suggestion
  selectMobileBadgeType(mobileBadgeType: MobileBadgeType): void {
```

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,304 @@
+  dropdownShown = false;
+  mobileDropdownShown = false;
+  mobileBadgeTypeDropdownShown = false;
+  showMobileTranslationBadges = true;
```
Can we make this an enum, e.g.

```suggestion
  enum MobileBadgeType {
    Question = "QUESTION",
    Translation = "TRANSLATION",
  }
```

It would make its usages clearer.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.spec.ts:
```diff
@@ -0,0 +1,350 @@
+          expect(component.mobileDropdownShown).toBeTrue();
+        }));
+
+      it('should should toggle mobile badge type dropdown', fakeAsync(() => {
```
```suggestion
      it('should toggle mobile badge type dropdown', fakeAsync(() => {
```

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,304 @@
+    const userContributionRights =
+      await this.userService.getUserContributionRightsDataAsync();
+    userContributionRights.can_review_translation_for_language_codes.map(
+      (code) => {
```
```suggestion
      (languageCode) => {
```

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,304 @@
+      contributionSubType: string, language?: string): Badge[] {
+    const badges: Badge[] = [];
+    for (const level of AppConstants.CONTRIBUTOR_BADGE_INITIAL_LEVELS) {
+      console.log(contributionCount);
```
Remove.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,304 @@
+        );
+      } else {
+        // From this block, we are adding locked badge(in predefined range)
+        // which will be achieved next.
```
```suggestion
        // Add a locked badge for the next unachieved milestone.
```

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/question-badge/question-badge.component.spec.ts:
```diff
@@ -0,0 +1,100 @@
+describe('Question badge component', () => {
+  let component: QuestionBadgeComponent;
+  let fixture: ComponentFixture<QuestionBadgeComponent>;
+  beforeEach(waitForAsync(() => {
+    TestBed.configureTestingModule({
+      imports: [HttpClientTestingModule],
+      declarations: [
+        QuestionBadgeComponent
+      ],
+      providers: [],
+      schemas: [NO_ERRORS_SCHEMA]
+    }).compileComponents();
+  }));
```
Add new line before and after.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/translation-badge/translation-badge.component.spec.ts:
```diff
@@ -0,0 +1,128 @@
+describe('Translation badge component', () => {
+  let component: TranslationBadgeComponent;
+  let fixture: ComponentFixture<TranslationBadgeComponent>;
+  beforeEach(waitForAsync(() => {
+    TestBed.configureTestingModule({
+      imports: [HttpClientTestingModule],
+      declarations: [
+        TranslationBadgeComponent
+      ],
+      providers: [],
+      schemas: [NO_ERRORS_SCHEMA]
+    }).compileComponents();
+  }));
```
Add new line before and after.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/question-badge/question-badge.component.spec.ts:
```diff
@@ -0,0 +1,100 @@
+    fixture.destroy();
+  });
+
+  describe('when a submission badge is passed ', () => {
```
What is meant by this describe?

chris7716 replied:
Updated the describe.

sagangwee commented on core/templates/pages/contributor-dashboard-page/question-badge/question-badge.component.spec.ts:
```diff
@@ -0,0 +1,100 @@
+
+      component.ngOnInit();
+
+      expect(component.contributionTypeText).toEqual('Submissions');
```
How does this test multiple badges?

chris7716 replied:
It should not actually test the multiple badges. It should test that the plural form of the badge text is given when the contributionCount is greater than 1.

sagangwee replied:
Got it. Thanks for updating!

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,304 @@
+    let level = 500;
+    while (contributionCount >= level) {
+      level += 500;
+      if (contributionCount >= level) {
+        badges.push(
+          {
+            contributionCount: level,
+            text: contributionSubType,
+            isUnlocked: true,
+            language
+          }
+        );
+      } else {
+        // From this block, we are adding locked badge(in non-predefined range)
+        // which will be achieved next.
+        badges.push(
+          {
+            contributionCount: level - contributionCount,
+            text: contributionSubType,
+            isUnlocked: false,
+            language
+          }
+        );
+        return badges;
```
Since this logic is the same as above, let's consolidate this into one loop (keep track of index, increment index, when index gets to end of `AppConstants.CONTRIBUTOR_BADGE_INITIAL_LEVELS`, start incrementing level by 500 instead).

chris7716 replied:
Done

sagangwee reviewed (changes requested):
@chris7716 Thanks took a pass.

chris7716 committed:
Address more comments

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Address comments

chris7716 reviewed (commented):
@sagangwee PTAL

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

chris7716 committed:
Fix checks

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,289 @@
+    let level = 0;
+    let index = 0;
+    while (contributionCount >= level) {
+      if (AppConstants.CONTRIBUTOR_BADGE_INITIAL_LEVELS.length > index) {
```
Nit:
```suggestion
      if (index < AppConstants.CONTRIBUTOR_BADGE_INITIAL_LEVELS.length) {
```
Just so that when reading you're thinking of the index first.

chris7716 replied:
Done

sagangwee commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,289 @@
+    this.dataLoading = false;
+  }
+
+  getObtainedBadges(
```
Looks great :)

sagangwee reviewed (approved):
@chris7716 Thanks, LGTM for badge-related changes.

BenHenning reviewed (dismissed):
@chris7716 is this stacked on #16289? It seems to include changes from that PR, but is out-of-date. The PR description doesn't mention this, so I'm a bit concerned that there's going to be a growing discrepancy in behavior between the base of this PR and #16289.

chris7716 committed:
Merge branch 'milestone2.2' of https://github.com/chris7716/oppia into milestone2.3
Update branch

chris7716 commented:
> @chris7716 is this stacked on #16289? It seems to include changes from that PR, but is out-of-date. The PR description doesn't mention this, so I'm a bit concerned that there's going to be a growing discrepancy in behavior between the base of this PR and #16289.

@BenHenning  Since this feature is also gated behind the feature I introduced in #16289 this is stacked on it. Feature gating logic is not changed here and now I have made this PR up-to-date with #16289. And updated the description as well.

BenHenning commented:
Thanks for confirming @chris7716. In that case, I don't think you'll need my approval since all changes in my codeowner files are already reflected in the earlier PR. Once that's merged & this is brought up-to-date with develop, GitHub should indicate my codeowners is no longer needed.

gp201 reviewed (approved):
Please fix the test and assign me again. Thanks!

gp201 reviewed (changes requested):
Please fix the test and assign me again. Thanks!

chris7716 committed:
Update branch[C

chris7716 committed:
Fix tests

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.html:
```diff
@@ -0,0 +1,552 @@
+<div class="stats-container">
+    <div class="badges-type-mobile-container">
+        <div class="dropdown-label">Badge Type</div>
+        <div #mobileBadgeTypeDropdown class="oppia-contribution-stats-type-selector-mobile-container e2e-test-topic-selector">
+            <div class="oppia-stats-type-selector-inner-container e2e-test-topic-selector-selected" (click)="toggleMobileBadgeTypeDropdown()">
+              {{mobileBadgeTypeSelected}}
+              <span class="material-icons oppia-contributions-dropdown-arrow" [ngStyle]="{ 'transform': mobileBadgeTypeSelected === 'Translation' ? 'scaleY(-1)' : 'scaleY(1)' }">
```
```suggestion
        <div #mobileBadgeTypeDropdown 
             class="oppia-contribution-stats-type-selector-mobile-container e2e-test-topic-selector">
            <div class="oppia-stats-type-selector-inner-container e2e-test-topic-selector-selected" 
                 (click)="toggleMobileBadgeTypeDropdown()">
              {{mobileBadgeTypeSelected}}
              <span class="material-icons oppia-contributions-dropdown-arrow"
                    [ngStyle]="{ 'transform': mobileBadgeTypeSelected === 'Translation' ? 'scaleY(-1)' : 'scaleY(1)' }">
```

ditto elsewhere

chris7716 replied:
Done

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.html:
```diff
@@ -0,0 +1,552 @@
+    top: 100%;
+    z-index: 1;
+  }
+  .oppia-stats-type-selector-dropdown-container::-webkit-scrollbar {
+    width: 7px;
+  }
+  .oppia-stats-type-selector-dropdown-container::-webkit-scrollbar-track {
+    border-radius: 5px;
+    box-shadow: inset 0 0 3px #808080;
+  }
+  .oppia-stats-type-selector-dropdown-container::-webkit-scrollbar-thumb {
+    background: #808080;
+    border-radius: 5px;
+    transition: background 0.5s ease;
+  }
+  .oppia-stats-type-selector-dropdown-container::-webkit-scrollbar-thumb:hover {
+    background: black;
+  }
```
Should we handle other than WebKit scrollbars too?

chris7716 replied:
Only these are handled throughout the CD page(selectors includin language selector, topic selector etc.). In order to keep the consistency, only these are handled in the badge component too.

vojtechjelinek replied:
Hmm, did you check that the UI works correctly in non-webkit browsers?

chris7716 replied:
I can not test it because I do not have a non-webkit browser. But should it be still an issue since we are already do this way in the contributor dashboard?

vojtechjelinek replied:
Can you ask someone for testing? Or can you download Firefox for testing?

chris7716 replied:
https://user-images.githubusercontent.com/41409951/203468683-e63721e3-5f5d-4e37-8f37-f4b147b420f4.mov

It seems fine in firefox.

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.html:
```diff
@@ -0,0 +1,552 @@
+    }
+  }
+</style>
+  
```
Missing newline?

chris7716 replied:
Done

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,294 @@
+import { ContributionAndReviewStatsService } from '../services/contribution-and-review-stats.service';
+
+interface ContributionCounts {
+  language?: string;
```
Why is language optional?

chris7716 replied:
We keep translation and question contribution counts in these objects. Only translation contributions have a language that is associated with it. Hence language is an optional attribute here.

vojtechjelinek replied:
Hmm, I think it would be better to use a bit different type then.

vojtechjelinek replied:
Or if you want to keep this as optinal then please do 
```suggestion
  language: string | null;
```
And set `null` explicitely.

The other possibility would be to use discriminated unions.

chris7716 replied:
Done

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,294 @@
+interface Badge {
+  contributionCount: number;
+  text: string;
+  language?: string;
```
Why is language optional?

chris7716 replied:
We keep translation and question contribution badges in these objects. Only translation contributions have a language that is associated with it. Hence language is an optional attribute here as well

vojtechjelinek replied:
ditto as above

chris7716 replied:
Done

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,294 @@
+  mobileDropdownShown = false;
+  mobileBadgeTypeDropdownShown = false;
+  mobileBadgeTypeSelected = MobileBadgeType.Translation;
+  selectedLanguage = '';
```
Why is this set to empty string and not null?

chris7716 replied:
Removed the assignment of empty string.

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,294 @@
+  }
+
+  getObtainedBadges(
+      contributionCount: number,
+      contributionSubType: string, language?: string): Badge[] {
```
```suggestion
      contributionCount: number,
      contributionSubType: string, 
      language?: string
  ): Badge[] {
```

Why is language optional?

chris7716 replied:
We create Question contribution badges from this function as well. They dont have an associated language. Hence it is an optional param here.

chris7716 replied:
Done

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/contributor-badge/contributor-badge.component.ts:
```diff
@@ -0,0 +1,294 @@
+  onDocumentClick(event: MouseEvent): void {
+    const targetElement = event.target as HTMLElement;
+    if (
+      targetElement &&
+       !this.dropdownRef.nativeElement.contains(targetElement)
```
```suggestion
      targetElement &&
      !this.dropdownRef.nativeElement.contains(targetElement)
```
ditto below

chris7716 replied:
Done

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/question-badge/question-badge.component.html:
```diff
@@ -0,0 +1,41 @@
+<div *ngIf="isUnlocked" class="badge-container">
+    <img class="translation-badge" src="/assets/images/contributor_dashboard/question-{{type}}-badge.component.svg">
```
```suggestion
    <img class="translation-badge" 
         src="/assets/images/contributor_dashboard/question-{{type}}-badge.component.svg">
```

chris7716 replied:
Done

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/question-badge/question-badge.component.html:
```diff
@@ -0,0 +1,41 @@
+     class="locked-badge-container"
+     container="body"
+     placement="bottom">
+    <img class="translation-badge" src="/assets/images/contributor_dashboard/locked-badge.svg">
```
```suggestion
  <img class="translation-badge" src="/assets/images/contributor_dashboard/locked-badge.svg">
```

chris7716 replied:
Done

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/question-badge/question-badge.component.html:
```diff
@@ -0,0 +1,41 @@
+        top: 70%;
+        font-size: 7px;
+    }
+</style>
```
Missing newline at the end of the file.

chris7716 replied:
Done

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/translation-badge/translation-badge.component.html:
```diff
@@ -0,0 +1,55 @@
+<div *ngIf="isUnlocked" class="badge-container">
```
I would expect that we have some kind of base badge object and then have a question and translation badge to inherit from that. Why isn't that so?

chris7716 replied:
We already have `Badge` object for both translation and question types. The only difference is translation type badges have language attribute.

We do not pass the whole badge object to the translation badge component or question badge component, instead we just pass their attributes.

vojtechjelinek replied:
I know. What I am saying is that it seems a bit redundant to have two full components if they do almost the same thing.

chris7716 replied:
Created a new component

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

chris7716 committed:
Update branch[C

chris7716 committed:
Fix fe tests

chris7716 committed:
Address changes

chris7716 committed:
Address changes

chris7716 committed:
Address changes

chris7716 committed:
Address changes

chris7716 committed:
Address changes

chris7716 committed:
Address changes

chris7716 committed:
Address changes

chris7716 committed:
Address changes

chris7716 committed:
Compress svgs

chris7716 reviewed (commented):
@vojtechjelinek PTAL

gp201 reviewed (approved):
LGTM

chris7716 committed:
Address changes

chris7716 committed:
Address changes

chris7716 committed:
Address changes

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/badge/badge.component.html:
```diff
@@ -0,0 +1,87 @@
+      {{contributionSubTypeText}}
+  </div>
+</div>
+<div *ngIf="isUnlocked && contributionType === 'translation'" class="translation-badge-container">
```
```suggestion
<div *ngIf="isUnlocked && contributionType === 'translation'" 
     class="translation-badge-container">
```

chris7716 replied:
Done

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/badge/badge.component.ts:
```diff
@@ -0,0 +1,71 @@
+  constructor() {}
+
+  ngOnInit(): void {
+    if (
+      this.contributionSubType === AppConstants
+        .CONTRIBUTION_STATS_SUBTYPE_SUBMISSION) {
+      this.contributionSubTypeText = 'Submission';
+    } else if (this.contributionSubType === AppConstants
+      .CONTRIBUTION_STATS_SUBTYPE_REVIEW) {
+      this.contributionSubTypeText = 'Review';
+    } else if (
+      this.contributionSubType === AppConstants
+        .CONTRIBUTION_STATS_SUBTYPE_CORRECTION
+    ) {
+      this.contributionSubTypeText = 'Correction';
+    } else {
```
You can implement mapping from the AppConstants.XXX → 'yyy' and then just do `mapping[XXX]` here to remove th ifs.

chris7716 replied:
Done

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/contributor-badges/contributor-badges.component.html:
```diff
@@ -0,0 +1,646 @@
+            </span>
+          </div>
+
+          <div *ngIf="dropdownShown" class="oppia-stats-type-selector-dropdown-container e2e-test-topic-selector-dropdown">
```
```suggestion
          <div *ngIf="dropdownShown" 
               class="oppia-stats-type-selector-dropdown-container e2e-test-topic-selector-dropdown">
```

chris7716 replied:
Done

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/contributor-badges/contributor-badges.component.html:
```diff
@@ -0,0 +1,646 @@
+        <div class="badge-box">
+          <div>Submissions</div>
+          <div class="badge-container">
+            <badge class="badge-item"
+                   *ngFor="let badge of translationBadges[selectedLanguage].submission"
+                   [isUnlocked]="badge.isUnlocked" contributionSubType="{{badge.text}}"
+                   language="{{badge.language}}"
+                   contributionType="translation" contributionCount="{{badge.contributionCount}}">
```
```suggestion
            <badge class="badge-item"
                   *ngFor="let badge of translationBadges[selectedLanguage].submission"
                   [isUnlocked]="badge.isUnlocked" 
                   contributionSubType="{{badge.text}}"
                   language="{{badge.language}}"
                   contributionType="translation" 
                   contributionCount="{{badge.contributionCount}}">
```

vojtechjelinek replied:
ditto below

chris7716 replied:
Done

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

chris7716 committed:
Address changes

chris7716 reviewed (commented):
@vojtechjelinek PTAL

chris7716 committed:
Fix errors

vojtechjelinek reviewed (approved):
LGTM!

chris7716 committed:
Fix errors

chris7716 committed:
Fix errors

chris7716 committed:
Fix errors
